gwpy/timeseries/timeseries.py

```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+            for the underlying heterodyne detection method
         """
+        # stride through the TimeSeries and heterodyne at a single frequency
+        phase = (2 * numpy.pi * f *
+                 self.dt.decompose().value *
+                 numpy.arange(0, self.size))
+        out = self.heterodyne(phase, stride=stride, singlesided=True)
+        if exp:
+            return out
+        mag = out.abs()
+        phase = type(mag)(numpy.angle(out, deg=deg))
+        phase.__array_finalize__(out)
+        phase.override_unit('deg' if deg else 'rad')
+        return (mag, phase)
+
+    def heterodyne(self, phase, stride=1, singlesided=False):
+        """Compute the average magnitude and phase of this `TimeSeries`
+        once per stride after heterodyning with a given phase series
+
+        Parameters
+        ----------
+        phase : `array_like`
+            an array of phase measurements (radians) with which to heterodyne
+            the signal
+
+        stride : `float`, optional
+            stride (seconds) between calculations, defaults to 1 second
+
+        singlesided : `bool`, optional
+            Boolean switch to return single-sided output (i.e., to multiply by
+            2 so that the signal is distributed across positive frequencies
+            only), default: False
+
+        Returns
+        -------
+        out : `TimeSeries`
+            magnitude and phase trends, represented as
+            `mag * exp(1j*phase)` with `dt=stride`
+
+        Notes
+        -----
+        This is similar to the :meth:`~gwpy.timeseries.TimeSeries.demodulate`
+        method, but is more general in that it accepts a varying phase
+        evolution, rather than a fixed frequency.
+
+        Unlike :meth:`~gwpy.timeseries.TimeSeries.demodulate`, the complex
+        output is double-sided by default, so is not multiplied by 2.
+
+        Examples
+        --------
+        Heterodyning can be useful in examining quasi-monochromatic signals
+        with a known phase evolution. Examples of such signals are the
+        `continuous-wave hardware injections
+        <https://www.gw-openscience.org/s6hwcw/>`_ in many of the LIGO
+        observing runs, which are designed to mimic the expected signals
+        from rapidly rotating neutron stars. For these sources the signal's
+        frequency slowly decreases and is also Doppler modulated due to
+        the Earth's rotational and orbital motion.
+
+        We can show an example of demodulation by creating a simulated signal
+        with a phase evolution described by a second order Taylor expansion in
+        phase, i.e., it is defined by a frequency and first frequency
+        derivative. We can download some LIGO Livingston data from GWOSC and
+        inject such a signal and attempt to recover the amplitude.
+
+        >>> from gwpy.timeseries import TimeSeries, TimeSeriesDict
+        >>> data = TimeSeries.fetch_open_data('L1', 1131350417, 1131357617)
```

Comment:
@mattpitkin, is there a reason so much data needs to be read in? Can you just read in the 180 seconds you need?

```suggestion
        >>> data = TimeSeries.fetch_open_data('L1', 1131350417, 1131350587)
```

Reply:
It doesn't need to be quite so much (I just copied from the demodulation example), but it does need to be enough to show the difference between the heterodyning and demodulation, which requires more than 180 seconds.

Reply:
I was working backwards through these, so have now seen you comment to remove the demodulation comparison.

Reply:
I've not changed it just 180 seconds, but have halved the amount of data.